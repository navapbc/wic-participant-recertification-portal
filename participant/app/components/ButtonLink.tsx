import type { ReactElement } from "react";
import { Link as RemixLink } from "@remix-run/react";
import type { LinkProps as RemixLinkProps } from "@remix-run/react";
import { Link as USWDSLink } from "@trussworks/react-uswds";
import type { CustomLinkProps as USWDSLinkProps } from "@trussworks/react-uswds/lib/components/Link/Link";

export type ButtonLinkProps = Omit<USWDSLinkProps<RemixLinkProps>, "asCustom">;

export const ButtonLink = (props: ButtonLinkProps): ReactElement => {
  const { className = "", variant = "unstyled", to, children, ...rest } = props;
  return (
    <USWDSLink
      asCustom={RemixLink}
      className={`usa-button ${className}`.trim()}
      variant={variant}
      to={to}
      {...rest}
    >
      {children}
    </USWDSLink>
  );
};
