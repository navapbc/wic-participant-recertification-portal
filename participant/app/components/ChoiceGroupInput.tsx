import { Trans } from "react-i18next";
import type { ChangeEvent, ReactElement } from "react";
import {
  Fieldset,
  Checkbox,
  Radio,
  ErrorMessage,
} from "@trussworks/react-uswds";

import Required from "app/components/Required";
import { useField } from "remix-validated-form";
import type { i18nKey, legendStyleType } from "app/types";

export type Choice = {
  value: string;
  labelElement: ReactElement;
};

export type ChoiceGroupInputProps = {
  name: string;
  choices: Choice[];
  legendKey: i18nKey;
  legendStyle?: legendStyleType;
  required?: boolean;
  handleChange?: (e: ChangeEvent<HTMLInputElement>) => void;
  type: "checkbox" | "radio";
  error?: string;
  helpElement?: ReactElement;
  keyBase: string;
};

export const ChoiceGroupInput = (
  props: ChoiceGroupInputProps
): ReactElement => {
  const {
    choices,
    legendKey,
    legendStyle = "default",
    required,
    type,
    helpElement,
    name,
    handleChange,
    keyBase,
  } = props;
  const { getInputProps, error } = useField(name);
  const InputTypeClass = type == "checkbox" ? Checkbox : Radio;
  if (!choices?.length) {
    return <></>;
  }
  const legendElement = (
    <div>
      <Trans i18nKey={legendKey} />
      {required ? <Required /> : ""}
    </div>
  );
  return (
    <>
      <Fieldset legend={legendElement} legendStyle={legendStyle} key={keyBase}>
        {error && (
          <ErrorMessage id={`${legendKey}-error-message`}>{error}</ErrorMessage>
        )}
        {helpElement}
        {choices?.map((choice: Choice) => (
          <InputTypeClass
            key={`${keyBase}-${choice.value}`}
            {...getInputProps({
              id: `${keyBase}-${choice.value}`,
              label: choice.labelElement,
              type: type,
              value: choice.value,
              onChange: handleChange,
            })}
          />
        ))}
      </Fieldset>
    </>
  );
};
