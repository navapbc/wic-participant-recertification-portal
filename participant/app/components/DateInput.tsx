import type { ReactElement } from "react";
import {
  DateInput as USWDSDateInput,
  Fieldset,
  DateInputGroup,
} from "@trussworks/react-uswds";
import Required from "app/components/Required";
import type { i18nKey, legendStyleType } from "~/types";
import { Trans, useTranslation } from "react-i18next";

type DateFieldTypes = "day" | "month" | "year";

export type DateInputProps = {
  id: string;
  name: string;
  dateKey: i18nKey;
  DMYorder?: boolean;
  legendKey: i18nKey;
  legendStyle?: legendStyleType;
  hint?: boolean;
  required?: boolean;
};

export const DateInput = (props: DateInputProps): ReactElement => {
  const {
    id,
    name,
    dateKey,
    legendKey,
    legendStyle = "default",
    hint = false,
    DMYorder = false,
    required = false,
  } = props;
  const { t } = useTranslation();
  const legendElement = (
    <div>
      <Trans i18nKey={legendKey} />
      {required ? <Required /> : ""}
    </div>
  );
  const hintKey = DMYorder ? `${dateKey}.hintDMY` : `${dateKey}.hintMDY`;
  const hintElement =
    hint && t(hintKey) ? (
      <div className="usa-hint" id="dateOfBirthHint">
        <Trans i18nKey={hintKey} />
      </div>
    ) : undefined;

  const orderedFields: DateFieldTypes[] = DMYorder
    ? ["day", "month", "year"]
    : ["month", "day", "year"];
  const orderedDateFields = orderedFields.map((field: DateFieldTypes) => {
    const len = field == "year" ? 4 : 2;
    return (
      <USWDSDateInput
        id={`${id}-${field}`}
        key={`${id}-${field}`}
        label={t(`${dateKey}.${field}`)}
        unit={field}
        maxLength={len}
        minLength={len}
        name={`${name}-${field}`}
      />
    );
  });
  return (
    <Fieldset legend={legendElement} legendStyle={legendStyle}>
      {hintElement}
      <DateInputGroup>{orderedDateFields}</DateInputGroup>
    </Fieldset>
  );
};
