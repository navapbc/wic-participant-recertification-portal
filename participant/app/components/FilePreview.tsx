import React, { MouseEventHandler, useEffect, useRef, useState } from "react";
import type { ReactElement } from "react";
import classNames from "classnames";
import { Button } from "@trussworks/react-uswds";
/** Moving the SPACER_GIF definition here instead of the constants.ts file,
 * as webpack was exporting that entire file, including use of the File
 * WebAPI; this was causing server-side site generators to break (#1250). */

const SPACER_GIF =
  "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7";

export type FilePreviewProps = {
  imageId: string;
  file: File;
  clickHandler: Function;
};

export const FilePreview = (props: FilePreviewProps): ReactElement => {
  const { imageId, file, clickHandler } = props;
  const fileReaderRef = useRef<FileReader>(new FileReader());
  const [isLoading, setIsLoading] = useState(true);
  const [previewSrc, setPreviewSrc] = useState(SPACER_GIF);
  const [showGenericPreview, setShowGenericPreview] = useState(false);

  useEffect(() => {
    fileReaderRef.current.onloadend = (): void => {
      setIsLoading(false);
      setPreviewSrc(fileReaderRef.current.result as string);
    };

    fileReaderRef.current.readAsDataURL(file);

    return (): void => {
      fileReaderRef.current.onloadend = null;
    };
  }, []);

  const { name } = file;

  const onImageError = (): void => {
    setPreviewSrc(SPACER_GIF);
    setShowGenericPreview(true);
  };

  const isPDF = name.indexOf(".pdf") > 0;
  const isWord = name.indexOf(".doc") > 0 || name.indexOf(".pages") > 0;
  const isVideo = name.indexOf(".mov") > 0 || name.indexOf(".mp4") > 0;
  const isExcel = name.indexOf(".xls") > 0 || name.indexOf(".numbers") > 0;
  const isGeneric = !isPDF && !isWord && !isVideo && !isExcel;

  const imageClasses = classNames("usa-file-input__preview-image", {
    "is-loading": isLoading,
    "usa-file-input__preview-image--pdf": showGenericPreview && isPDF,
    "usa-file-input__preview-image--word": showGenericPreview && isWord,
    "usa-file-input__preview-image--video": showGenericPreview && isVideo,
    "usa-file-input__preview-image--excel": showGenericPreview && isExcel,
    "usa-file-input__preview-image--generic": showGenericPreview && isGeneric,
  });

  return (
    <div className="margin-bottom-4 margin-top-2">
      <div className="usa-file-input">
        <div className="usa-file-input__preview-heading border-bottom-width-0 border-width-1px border-dashed">
          <div>1 file selected</div>
          <Button
            type="button"
            unstyled
            className="usa-file-input__choose text-secondary-vivid z-top"
            onClick={() => clickHandler(name)}
          >
            Remove File
          </Button>
        </div>

        <div
          className="usa-file-input__preview border-dashed border-top-width-0 border-width-1px"
          aria-hidden="true"
        >
          <img
            id={imageId}
            data-testid="file-input-preview-image"
            src={previewSrc}
            alt=""
            className={imageClasses}
            onError={onImageError}
          />
          {name}
        </div>
      </div>
    </div>
  );
};
