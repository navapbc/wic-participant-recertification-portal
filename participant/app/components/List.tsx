import type { ReactElement } from "react";
import type { i18nKey } from "app/types";
import { Trans } from "react-i18next";

export type ListProps = {
  listKeys: i18nKey[];
  ordered?: boolean;
  unstyled?: boolean;
};

export const List = (props: ListProps): ReactElement => {
  const { listKeys, ordered = false, unstyled = false } = props;
  const classNames = `usa-list ${unstyled ? "usa-list--unstyled" : ""}`;
  const ListTag = ordered ? "ol" : "ul";

  return (
    <ListTag className={classNames.trim()}>
      {listKeys.map((key: string) => (
        <li key={key}>
          <Trans i18nKey={key} />
        </li>
      ))}
    </ListTag>
  );
};

export default List;
