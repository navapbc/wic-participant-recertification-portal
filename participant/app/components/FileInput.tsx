import { FormGroup, Label } from "@trussworks/react-uswds";
import { FilePreview } from "app/components/internal/FilePreview";
import { FileInputControl } from "app/components/internal/FileInputControl";
import React, {
  useState,
  useRef,
  useImperativeHandle,
  useEffect,
  forwardRef,
} from "react";
import { i18nKey } from "~/types";
import { Trans, useTranslation } from "react-i18next";

export type FileInputProps = {
  className?: string;
  id: string;
  name: string;
  labelKey: i18nKey;
  accept?: string;
  required?: boolean;
  maxFileSizeInBytes?: number;
  maxFileCount?: number;
};

export type FileState = {
  [key: string]: File;
};

export type FileInputRef = {
  input: HTMLInputElement | null;
  files: File[];
};

export type FileError = {
  errorType?: "count" | "size" | "type";
};

export const FileInputForwardRef: React.ForwardRefRenderFunction<
  FileInputRef,
  FileInputProps & JSX.IntrinsicElements["input"]
> = (
  {
    className,
    id,
    name,
    labelKey,
    accept,
    required,
    maxFileSizeInBytes = 26_214_400,
    maxFileCount = 25,
    ...inputProps
  },
  ref
): React.ReactElement => {
  const { t } = useTranslation();
  const [files, setFiles] = useState({} as FileState);
  const [previews, setPreviews] = useState(<></>);
  const fileInputRef = useRef(null);
  const documentString = t(`${labelKey}.document`);
  const [errorMessage, setErrorMessage] = useState("");

  const ifNeeded = t(`${labelKey}.ifNeeded`);
  const currentDocumentNumber = Object.keys(files).length;

  useImperativeHandle(
    ref,
    () => ({
      input: fileInputRef.current,
      files: Object.keys(files).map((key) => files[key]),
    }),
    [files]
  );
  const makeSafeForID = (name: string): string => {
    return name.replace(/[^a-z0-9]/g, function replaceName(s) {
      const c = s.charCodeAt(0);
      if (c === 32) return "-";
      if (c >= 65 && c <= 90) return `img_${s.toLowerCase()}`;
      return `__${c.toString(16).slice(-4)}`;
    });
  };

  const renderPreviews = () => {
    setPreviews(
      <>
        {Object.keys(files).map((fileName, index) => {
          let file = files[fileName];
          const imageId = makeSafeForID(fileName);
          return (
            <div key={`document-${index + 1}`}>
              <span className="font-sans-lg">{`${documentString} ${
                index + 1
              }`}</span>
              <FilePreview
                imageId={`preview-${index}`}
                file={file}
                clickHandler={removeFile}
                removeFileKey={`${labelKey}.removeFile`}
                selectedKey={`${labelKey}.selected`}
                altTextKey={`${labelKey}.altText`}
              />
            </div>
          );
        })}
      </>
    );
  };
  useEffect(renderPreviews, [files]);
  const fileTypeCheck = (newFile: File): boolean => {
    if (!accept || accept == "*") {
      return true;
    }
    const acceptedTypes = accept.split(",");
    let fileTypeAllowed = true;
    if (fileTypeAllowed) {
      for (let j = 0; j < acceptedTypes.length; j += 1) {
        const fileType = acceptedTypes[parseInt(`${j}`)];
        fileTypeAllowed =
          newFile.name.indexOf(fileType) > 0 ||
          newFile.type.includes(fileType.replace(/\*/g, ""));
        if (fileTypeAllowed) break;
      }
    }
    return fileTypeAllowed;
  };
  const addNewFiles = (newFiles: FileList) => {
    setErrorMessage("");

    if (
      Object.keys(files).length + Object.keys(newFiles).length <=
      maxFileCount
    ) {
      for (let file of newFiles) {
        if (file.size <= maxFileSizeInBytes) {
          if (fileTypeCheck(file)) {
            files[file.name] = file;
          } else {
            setErrorMessage(`${labelKey}.fileTypeError`);
          }
        } else {
          setErrorMessage(`${labelKey}.fileSizeError`);
        }
      }
    } else {
      setErrorMessage(`${labelKey}.fileCountError`);
    }
    return { ...files };
  };
  const handleNewFileUpload = (e: React.ChangeEvent<HTMLInputElement>) => {
    setErrorMessage("");
    const { files: newFiles } = e.target;
    if (newFiles?.length) {
      let updatedFiles = addNewFiles(newFiles);
      setFiles(updatedFiles);
    }
  };
  const removeFile = (fileName: string) => {
    setErrorMessage("");
    delete files[fileName];
    setFiles({ ...files });
  };

  const currentDocumentString =
    currentDocumentNumber > 0
      ? `${documentString} ${currentDocumentNumber + 1} ${ifNeeded}`
      : `${documentString} ${currentDocumentNumber + 1}`;
  return (
    <FormGroup>
      {previews}
      <Label htmlFor="file-input-async" className="font-sans-lg">
        {currentDocumentString}
      </Label>
      {errorMessage && (
        <div data-testid="file-input-error" className="file-input-error">
          <Trans i18nKey={errorMessage} />
        </div>
      )}
      <FileInputControl
        {...inputProps}
        ref={fileInputRef}
        id={id}
        name={name}
        accept={accept}
        onChange={handleNewFileUpload}
        onDrop={function noRefCheck() {}}
        value=""
        title=""
        draggable={true}
        emptyKey={`${labelKey}.noFiles`}
        notEmptyKey={`${labelKey}.additionalFiles`}
        fileTypeErrorKey={`${labelKey}.fileTypeError`}
        empty={currentDocumentNumber == 0}
      />
    </FormGroup>
  );
};

export const FileInput = forwardRef(FileInputForwardRef);
