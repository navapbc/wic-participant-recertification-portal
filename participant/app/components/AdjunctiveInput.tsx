import type { i18nKey } from "~/types";
import type { ReactElement } from "react";
import { Trans, useTranslation } from "react-i18next";
import { ChoiceGroupInput } from "~/components/ChoiceGroupInput";
import { List } from "app/components/List";
import type {
  Choice,
  ChoiceGroupInputProps,
} from "~/components/ChoiceGroupInput";

export type AdjunctiveInputProps = Omit<
  ChoiceGroupInputProps,
  "choices" | "type" | "legendKey"
> & {
  adjunctiveKey: i18nKey;
};

export const AdjunctiveInput = (props: AdjunctiveInputProps): ReactElement => {
  const { t } = useTranslation();

  const { name, adjunctiveKey, legendStyle, ...rest } = props;
  const adjunctiveChoices: Choice[] = [
    {
      value: "yes",
      labelElement: <Trans i18nKey={`${adjunctiveKey}.yes`} />,
    },
    {
      value: "no",
      labelElement: <Trans i18nKey={`${adjunctiveKey}.no`} />,
    },
  ];
  // eslint-disable-next-line  @typescript-eslint/no-unnecessary-type-assertion
  const programList = t(`${adjunctiveKey}.programs`, {
    returnObjects: true,
  }) as Array<string>;
  const helpElement = (
    <List listKeys={programList} ordered={false} />
  );
  return (
    <ChoiceGroupInput
      choices={adjunctiveChoices}
      name={name}
      type="radio"
      legendKey={`${adjunctiveKey}.label`}
      legendStyle={legendStyle}
      helpElement={helpElement}
      {...rest}
    />
  );
};
