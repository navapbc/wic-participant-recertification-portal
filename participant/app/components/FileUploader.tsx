import { FormGroup, Label } from "@trussworks/react-uswds";
import { FilePreview } from "app/components/FilePreview";
import { FileInput } from "app/components/internal/FileInput";
import { Trans, useTranslation } from "react-i18next";
import React, {
  useState,
  useRef,
  useImperativeHandle,
  useEffect,
  forwardRef,
} from "react";
import type { ReactElement } from "react";
import type { PreviousUpload, i18nKey } from "~/types";
import type { FileInputRef } from "app/components/internal/FileInput";

export type FileUploaderProps = {
  className?: string;
  id: string;
  name: string;
  labelKey: i18nKey;
  accept?: string;
  required?: boolean;
  maxFileSizeInBytes?: number;
  maxFileCount?: number;
  children?: ReactElement;
};

export type FileUploaderRef = {
  files: File[];
  removeFileList: (removeList: PreviousUpload[]) => void;
};

export type FileState = {
  [key: string]: File;
};

export type FileError = {
  errorType?: "count" | "size" | "type";
};

export const FileUploaderForwardRef: React.ForwardRefRenderFunction<
  FileUploaderRef,
  FileUploaderProps & JSX.IntrinsicElements["input"]
> = (
  {
    className,
    id,
    name,
    labelKey,
    accept,
    required,
    maxFileSizeInBytes = 26_214_400,
    maxFileCount = 20,
    children,
    ...inputProps
  },
  ref
): React.ReactElement => {
  const { t } = useTranslation();
  const [files, setFiles] = useState<FileState>({});
  const [previews, setPreviews] = useState(<></>);
  const fileInputRef = useRef<FileInputRef>(null);
  const documentString = t(`${labelKey}.document`);
  const [errorMessage, setErrorMessage] = useState("");

  const ifNeeded = t(`${labelKey}.ifNeeded`);
  const currentDocumentNumber = Object.keys(files).length;

  /* This allows us to add callable methods and values to a reference
   * passed in by the rendering page, instead of relying on a callback
   * passed in on the page to pass objects held in state (in this case, files)
   * out to the parent.
   * We're also adding a ref directly to the input component rendered here -
   * so if we need to manage its state or settings, that can be done.
   * For more, see: https://blog.webdevsimplified.com/2022-06/use-imperative-handle/
   */
  useImperativeHandle(
    ref,
    () => ({
      files: Object.keys(files).map((key) => files[key]),
      removeFileList: removeFileList,
    }),
    // eslint-disable-next-line   react-hooks/exhaustive-deps -- (deps list is correct)
    [files]
  );

  const renderPreviews = () => {
    setPreviews(
      <>
        {Object.keys(files).map((fileName, index) => {
          let file = files[fileName];
          return (
            <div
              key={`document-${index + 1}`}
              className={index == 0 ? "margin-top-2" : undefined}
            >
              <span>{`${documentString} ${index + 1}`}</span>
              <FilePreview
                imageId={`preview-${index}`}
                file={file}
                name={file.name}
                clickHandler={removeFile}
                removeFileKey={`${labelKey}.removeFile`}
                selectedKey={`${labelKey}.selected`}
                altTextKey={`${labelKey}.altText`}
              />
            </div>
          );
        })}
      </>
    );
  };
  // eslint-disable-next-line   react-hooks/exhaustive-deps -- (deps list is correct)
  useEffect(renderPreviews, [files, documentString, labelKey]);
  const fileTypeCheck = (newFile: File): boolean => {
    if (!accept || accept == "*") {
      return true;
    }
    const acceptedTypes = accept.split(",");
    let fileTypeAllowed = true;
    for (let j = 0; j < acceptedTypes.length; j += 1) {
      const fileType = acceptedTypes[parseInt(`${j}`)];
      fileTypeAllowed =
        newFile.name.indexOf(fileType) > 0 ||
        newFile.type.includes(fileType.replace(/\*/g, ""));
      if (fileTypeAllowed) break;
    }

    return fileTypeAllowed;
  };
  const addNewFiles = (newFiles: FileList) => {
    setErrorMessage("");

    if (
      Object.keys(files).length + Object.keys(newFiles).length <=
      maxFileCount
    ) {
      for (let file of newFiles) {
        if (file.size <= maxFileSizeInBytes) {
          if (fileTypeCheck(file)) {
            files[file.name] = file;
          } else {
            setErrorMessage(`${labelKey}.fileTypeError`);
          }
        } else {
          setErrorMessage(`${labelKey}.fileSizeError`);
        }
      }
    } else {
      setErrorMessage(`${labelKey}.fileCountError`);
    }
    return { ...files };
  };
  const removeFileList = (removeList: PreviousUpload[]) => {
    setErrorMessage("");
    setFiles({});
    for (let file of removeList) {
      delete files[file.name];
      setFiles({ ...files });
    }
  };
  const handleNewFileUpload = (e: React.ChangeEvent<HTMLInputElement>) => {
    setErrorMessage("");
    const { files: newFiles } = e.target;
    if (newFiles?.length) {
      let updatedFiles = addNewFiles(newFiles);
      setFiles(updatedFiles);
    }
  };
  const removeFile = (fileName: string) => {
    setErrorMessage("");
    delete files[fileName];
    setFiles({ ...files });
  };

  const currentDocumentString =
    currentDocumentNumber > 0
      ? `${documentString} ${currentDocumentNumber + 1} ${ifNeeded}`
      : `${documentString} ${currentDocumentNumber + 1}`;
  return (
    <FormGroup>
      {children}
      {previews}
      <Label htmlFor={id}>{currentDocumentString}</Label>
      {errorMessage && (
        <div data-testid="file-input-error" className="file-input-error">
          <span className="display-block">
            <Trans i18nKey={errorMessage} />
          </span>
        </div>
      )}
      <FileInput
        {...inputProps}
        ref={fileInputRef}
        id={id}
        name={name}
        accept={accept}
        onChange={handleNewFileUpload}
        onDrop={function noRefCheck() {}}
        value=""
        title=""
        draggable={true}
        emptyKey={`${labelKey}.noFiles`}
        notEmptyKey={`${labelKey}.additionalFiles`}
        fileTypeErrorKey={`${labelKey}.fileTypeError`}
        empty={currentDocumentNumber == 0}
      />
      <noscript>
        <Trans i18nKey={`${labelKey}.noScriptMessage`} />
      </noscript>
    </FormGroup>
  );
};

export const FileUploader = forwardRef(FileUploaderForwardRef);
