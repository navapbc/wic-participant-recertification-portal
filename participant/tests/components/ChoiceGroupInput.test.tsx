import { screen } from "@testing-library/react";
import { renderWithRouter } from "tests/helpers/setup";

import { ChoiceGroupInput } from "app/components/ChoiceGroupInput";

import type {
  ChoiceGroupInputProps,
  Choice,
} from "app/components/ChoiceGroupInput";
import { useField } from "remix-validated-form";
jest.mock("remix-validated-form");
import { mockUseField } from "tests/helpers/remixValidatedFormMock";
const mockedUseField = jest.mocked(useField);
mockedUseField.mockImplementation(mockUseField);

const choices: Choice[] = ["a", "b", "c"].map((option) => {
  return {
    value: option,
    labelElement: <div>label {option}</div>,
  };
});

const testProps: Omit<ChoiceGroupInputProps, "type"> = {
  choices: choices,
  name: "test-choices",
  legendKey: "title",
  required: false,
};

it("should match snapshot when it is a set of checkboxes", () => {
  const { container } = renderWithRouter(
    <ChoiceGroupInput {...testProps} type="checkbox" />
  );
  expect(container).toMatchSnapshot();
});

it("should match snapshot when it is a set of radio buttons", () => {
  const { container } = renderWithRouter(
    <ChoiceGroupInput {...testProps} type="radio" />
  );
  expect(container).toMatchSnapshot();
});

it("should match display required marker if required is true", () => {
  renderWithRouter(
    <ChoiceGroupInput {...testProps} required={true} type="checkbox" />
  );
  const required = screen.getByText("*");
  expect(required).toBeInTheDocument;
});

it("should display an helpElement if passed as props", () => {
  renderWithRouter(
    <ChoiceGroupInput
      {...testProps}
      type="checkbox"
      helpElement={<div data-testid="test-help">Helpful Help</div>}
    />
  );
  const helpElement = screen.getByTestId("test-help");
  expect(helpElement).toBeInTheDocument;
  expect(helpElement.textContent).toBe("Helpful Help");
});
