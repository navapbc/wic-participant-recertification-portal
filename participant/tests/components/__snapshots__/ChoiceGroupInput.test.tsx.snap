// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot when it is a set of checkboxes 1`] = `
<div>
  <fieldset
    class="usa-fieldset"
    data-testid="fieldset"
  >
    <legend
      class="usa-legend"
    >
      <div>
        title
      </div>
    </legend>
    <div
      class="usa-checkbox"
      data-testid="checkbox"
    >
      <input
        class="usa-checkbox__input"
        id="test-choices-a"
        name="test-choices"
        type="checkbox"
        value="a"
      />
      <label
        class="usa-checkbox__label"
        for="test-choices-a"
      >
        <div>
          label 
          a
        </div>
      </label>
    </div>
    <div
      class="usa-checkbox"
      data-testid="checkbox"
    >
      <input
        class="usa-checkbox__input"
        id="test-choices-b"
        name="test-choices"
        type="checkbox"
        value="b"
      />
      <label
        class="usa-checkbox__label"
        for="test-choices-b"
      >
        <div>
          label 
          b
        </div>
      </label>
    </div>
    <div
      class="usa-checkbox"
      data-testid="checkbox"
    >
      <input
        class="usa-checkbox__input"
        id="test-choices-c"
        name="test-choices"
        type="checkbox"
        value="c"
      />
      <label
        class="usa-checkbox__label"
        for="test-choices-c"
      >
        <div>
          label 
          c
        </div>
      </label>
    </div>
  </fieldset>
</div>
`;

exports[`should match snapshot when it is a set of radio buttons 1`] = `
<div>
  <fieldset
    class="usa-fieldset"
    data-testid="fieldset"
  >
    <legend
      class="usa-legend"
    >
      <div>
        title
      </div>
    </legend>
    <div
      class="usa-radio"
      data-testid="radio"
    >
      <input
        class="usa-radio__input"
        id="test-choices-a"
        name="test-choices"
        type="radio"
        value="a"
      />
      <label
        class="usa-radio__label"
        for="test-choices-a"
      >
        <div>
          label 
          a
        </div>
      </label>
    </div>
    <div
      class="usa-radio"
      data-testid="radio"
    >
      <input
        class="usa-radio__input"
        id="test-choices-b"
        name="test-choices"
        type="radio"
        value="b"
      />
      <label
        class="usa-radio__label"
        for="test-choices-b"
      >
        <div>
          label 
          b
        </div>
      </label>
    </div>
    <div
      class="usa-radio"
      data-testid="radio"
    >
      <input
        class="usa-radio__input"
        id="test-choices-c"
        name="test-choices"
        type="radio"
        value="c"
      />
      <label
        class="usa-radio__label"
        for="test-choices-c"
      >
        <div>
          label 
          c
        </div>
      </label>
    </div>
  </fieldset>
</div>
`;
